content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,562 @@
+
+lstm_network = nn.LSTM(input_size=dimension,
+                       hidden_size=dimension,
+                       num_layers=num_layer,
+                       bidirectional=bidirectional,
+                       batch_first=True)
+
+input_dim = dimension * 2 #since bidirectional
+
+fully_connected_network = nn.Linear(input_dim, tag_dimension)
+
+full_network_bi_lstm = FullNetWork(lstm_network, fully_connected_network)
+```
+ 
+### Training
+ 
+ ```python
+exp_bi_lstm = Experiment("./", full_network_bi_lstm, device=cuda_device, optimizer=optimizer,
+                 loss_function=cross_entropy, batch_metrics=["acc"])
+exp_bi_lstm.train(train_loader, valid_generator=valid_loader, epochs=epoch_number)
+```
+
+### Results
+Here the last epoch validation results of the larger model. On the validation dataset,
+we can see that we obtain a marginal gain of around `0.3` for the accuracy over the previous one. Not much of an improvement.
+
+|   Model  | Bi-LSTM two layers |
+|:--------:|:------------------:|
+|   Loss   |    0.0050          |
+| Accuracy |    99.8594         |
+
+But now that we have our two trained models, let's use the test set as a final an **unique** steps of the evaluation of the performance.
+
+``` python
+exp.test(test_loader)
+exp_bi_lstm.test(test_loader)
+```
+
+The next table presents the results of the Bi-LSTM with two layers and the previous model (LSTM with one layer).
+
+|   Model  | LSTM one layer | Bi-LSTM two layers |
+|:--------:|:--------------:|:------------------:|
+|   Loss   |     0.0152     |    **0.0050**      |
+| Accuracy |     99.5758    |    **99.8550**     |
+
+We can see similar results as the validation one for both the model. Also, we still see a little improvement of the accuracy and the loss for the larger model. But when considering that we only improved by around 0.3, one can
+argue that it's only a matter of the seed. For testing our approach's robustness, we could use retrain multiple times
+our training step but using every time a different seed. Using those trained models, we can report the mean and one standard variation of the metrics instead of a single training. But instead of doing that, let's try something else. 
+
+#### Zero Shot Evaluation
+Since we have to our disposition other country address, let's see if our model has really learned a typical address sequence
+or he has simply learned all the cases (know as memorization).
+
+We will test three different cases
+
+    - the first one we will test if our model performs well on countries using the exact same 
+    address structure as our training dataset: United-States of America (US) and United-Kingdom (UK)
+    - the second one we will test if our model performs well on a country using the exact same address structure **but** 
+    using a totally different language: Russia (RU)
+    - the last one will test if our model performs well on a country using a different address structure **and** a different language: Mexico (MX).
```

Comment:
It seems that the bullet points do not appear in the markdown. I think the spaces before them should be removed.

Reply:
cf90662